src/garbage.rs

```diff
@@ -0,0 +1,140 @@
+//! Garbage collection.
+//!
+//! # Garbages
+//!
+//! Objects that get unlinked from concurrent data structures must be stashed away until the global
+//! epoch sufficiently advances so that they become safe for destruction.  We call these object
+//! garbages.  When the global epoch advances sufficiently, `Destroy` garbages are dropped (i.e. the
+//! destructors are called), and `Free` garbages are freed.  In addition, you can register arbitrary
+//! function to be called later using the `Fn` garbages.
+//!
+//! # Bags
+//!
+//! Pointers to such garbages are pushed into thread-local bags, and when it becomes full, the bag
+//! is marked with the current global epoch and pushed into a global queue of garbage bags.  We
+//! store garbages in thread-local storages for amortizing the synchronization cost of pushing the
+//! garbages to a global queue.
+//!
+//! # Garbage queues
+//!
+//! Whenever a bag is pushed into a queue, some garbage in the queue is collected and destroyed
+//! along the way.  Garbage collection can also be manually triggered by calling `collect()`.  This
+//! design reduces contention on data structures.  Ideally each instance of concurrent data
+//! structure may have it's own queue that gets fully destroyed as soon as the data structure gets
+//! dropped.
+
+use std::mem;
+use boxfnonce::SendBoxFnOnce;
+use arrayvec::ArrayVec;
+
+/// Maximum number of objects a bag can contain.
+#[cfg(not(feature = "strict_gc"))]
+const MAX_OBJECTS: usize = 64;
+#[cfg(feature = "strict_gc")]
+const MAX_OBJECTS: usize = 4;
+
+
+pub enum Garbage {
+    Destroy {
+        object: *mut u8,
+        size: usize,
+        destroy: unsafe fn(*mut u8, usize),
+    },
+    Free { object: *mut u8, size: usize },
+    Fn { f: Option<SendBoxFnOnce<(), ()>> },
+}
+
+unsafe impl Send for Garbage {}
+
+impl Garbage {
+    /// Make a garbage object that will later be destroyed using `destroy`.
+    ///
+    /// The specified object is an array allocated at address `object` and consists of `size`
+    /// elements of type `T`.
+    ///
+    /// Note: The object must be `Send + 'self`.
+    pub fn new_destroy<T>(object: *mut T, size: usize, destroy: unsafe fn(*mut T, usize)) -> Self {
+        Garbage::Destroy {
+            object: object as *mut u8,
+            size: size,
+            destroy: unsafe { mem::transmute(destroy) },
```

Comment:
Pointers to different data types are _not_ guaranteed to be of the same size (even though they usually are). Hence, transmuting `fn(*mut T, usize)` to `fn(*mut u8, usize)` might produce wrong results.